### ISSUE TITLE
USB WWAN network-manager connection fails to start

### ISSUE THREAD

**Issue by evax** [NixOS/nixpkgs#72330]
**Describe the bug**
The network-manager fails to start a previously working 3G network connection using an USB dongle since the update to 19.09.

**To Reproduce**
1. Connect the dongle
2. Wait for the connection to appear in the network-manager list
3. Try to initiate the connection

**Expected behavior**
The connection should start

**Additional context**
The connection fails to start with the following error: `error starting PPP: Could not find "pppd" binary`

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.3.7, NixOS, 19.09.976.c75de8bc12c (Loris)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3`
 - channels(thomas): `"unstable-20.03pre198214.4cd2cb43fb3"`
 - channels(root): `"nixos-19.09.976.c75de8bc12c"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by worldofpeace** [comment]
This is `19.03` to `19.09` correct?

**Comment by worldofpeace** [comment]
This error is interesting, we give an absolute path for `pppd` with `-Dpppd=${ppp}/bin/pppd` in meson.
Perhaps this suggests that had no effect.

**Comment by evax** [comment]
Yes, `19.03` to `19.09`

**Comment by evax** [comment]
Stracting the NetworkManager process:
```
stat("/dev/ppp", {st_mode=S_IFCHR|0600, st_rdev=makedev(0x6c, 0), ...}) = 0
access("/nix/store/kwzfflbp2idaa9g94h86i3al6fkkvi92-network-manager-1.18.4/sbin/pppd", X_OK) = -1 ENOENT (No such file or directory)
access("/nix/store/kwzfflbp2idaa9g94h86i3al6fkkvi92-network-manager-1.18.4/bin/pppd", X_OK) = -1 ENOENT (No such file or directory)
access("/usr/local/sbin/pppd", X_OK)    = -1 ENOENT (No such file or directory)
access("/sbin/pppd", X_OK)              = -1 ENOENT (No such file or directory)
access("/usr/sbin/pppd", X_OK)          = -1 ENOENT (No such file or directory)
access("/usr/local/bin/pppd", X_OK)     = -1 ENOENT (No such file or directory)
access("/bin/pppd", X_OK)               = -1 ENOENT (No such file or directory)
access("/usr/bin/pppd", X_OK)           = -1 ENOENT (No such file or directory)
```

So I think I have an easy workaround putting pppd in one of the well known locations, but still something is wrong with the build.

**Comment by worldofpeace** [comment]
Here's what we figured out on IRC

> @evax 
>  I updated the ticket with some strace - it's looking in the wrong place it seems
>
> @worldofpeace 
>  it is indeed, and nothing is wrong with our build config. What's weird is we hardcode binaries the same way for everything and they work, but just for pppd it doesn't care about the build substituted one.
>  in strace, are there ENOENT for other binaries?
> 
> @evax 
> not in the trace I have (but I went straight to testing this)
> I did a ln -s `which pppd` /bin/pppd for now and I'm up and running
> 
> @worldofpeace 
> !!! how frightening
>
> @worldofpeace 
>  PPPD_PATH is missing in https://gitlab.freedesktop.org/NetworkManager/NetworkManager/blob/master/config.h.meson so it is always null
> I think I can make a patch upstream, and maybe forward you branch to test it with
> 
> @evax 
>  this is 1.18.4 though
>
> @worldofpeace 
>  I can make a branch variant for stable, they'll likely backport my patch anyway.
>
> @evax 
> ah I see, great, thanks a lot!

Will be doing what I promised :+1:

**Comment by worldofpeace** [comment]
@evax You can test with the following branch on 19.09
* https://github.com/worldofpeace/nixpkgs/commits/test-pppd-path

**Comment by worldofpeace** [comment]
PRs submitted:
* https://github.com/NixOS/nixpkgs/pull/72362
* https://github.com/NixOS/nixpkgs/pull/72363

patch upstream:
* https://gitlab.freedesktop.org/NetworkManager/NetworkManager/merge_requests/323

**Comment by evax** [comment]
This is fixed in the latest 19.09 update, thanks @worldofpeace for an incredibly fast diagnosis and fix.
